src/lightning/__init__.py

```diff
@@ -15,21 +16,28 @@
 
 from lightning.__about__ import *  # noqa: E402, F403
 from lightning.__version__ import version as __version__  # noqa: E402
-from lightning.fabric.fabric import Fabric  # noqa: E402
-from lightning.fabric.utilities.seed import seed_everything  # noqa: E402
-from lightning.pytorch.callbacks import Callback  # noqa: E402
-from lightning.pytorch.core import LightningDataModule, LightningModule  # noqa: E402
-from lightning.pytorch.trainer import Trainer  # noqa: E402
-
-__all__ = [
-    "Trainer",
-    "LightningDataModule",
-    "LightningModule",
-    "Callback",
-    "seed_everything",
-    "Fabric",
-    "__version__",
-]
+
+# This enables us to control imports in a more granular way for performance reasons
+lazy_import = bool(int(os.getenv("LIGHTNING_LAZY_IMPORTS", "0")))
+
+if not lazy_import:
```

Comment:
If the environment sets this variable and the user doesn't know it, it's going to be ImportError hell.

In that case you should do something like


```python
if not lazy_import:
    ...
else:
    def __getattr__(name):
        raise RuntimeError("Lazy imports are enabled ...")
```

Also please test this, it can be finicky and do weird stuff with imports: https://stackoverflow.com/questions/60774515/python-getattr-executed-multiple-times/60803436#60803436

Reply:
This would never be exposed to the user, so there is no risk there. I don't think we need extra checks there.